samples/bot-sso-adaptivecard/nodejs/bots/botSSOAdaptiveCard.js

```diff
@@ -0,0 +1,171 @@
+
+        this.onMessage(async (context, next) => {
+            const text = context.activity.text;
+
+            if (text) {
+                switch (text) {
+                    case "login":
+                        await context.sendActivity({ attachments: [this.getInitialAdaptiveCard(),] });
+                        break;
+
+                    case "PerformSSO":
+                        await context.sendActivity({ attachments: [this.getAdaptiveCardWithSSO()] });
+                        break;
+
+                    default:
+                        const message = "Please send 'login' for options";
+                        await context.sendActivity(message);
+                        break;
+                }
+            }
+            // By calling next() you ensure that the next BotHandler is run.
+            await next();
+        });
+    }
+
+    /**
+    * Invokes when Action.Submit button performs onclick action. 
+    * @param {turnContext} context
+    */
+    async onInvokeActivity(context) {
+        if (context._activity.name == "adaptiveCard/action") {
+            if (context._activity.value == null) {
+                return null;
+            }
+
+            var value = context._activity.value;
+
+            if (value["action"] == null) {
+                return null;
+            }
+
+            var actiondata = value["action"];
+
+            if (actiondata["verb"] == null) {
+                return null;
+            }
+
+            // authToken and state are absent, handle verb
+            let verb = actiondata["verb"];
+
+            var authentication = null;
+
+            // When adaptiveCard / action invoke activity from teams contains token in response to sso flow from earlier invoke.
+            if (value["authentication"] != null) {
+                authentication = value["authentication"];
+            }
+
+            if (authentication == null) {
+                // authToken and state are absent, handle verb
+                switch (verb) {
+                    case "initiateSSO":
+                        //when token is absent in the invoke.
+                        return this.getInitialAdaptiveCard();
+                }
+            }
+            else {
+                // when token is present, send authentication token.
+                return this.createAdaptiveCardInvokeResponseAsync(authentication, context);
+            }
+        }
+
+        return null;
+    }
+
+    /**
+    * AuthToken present. Verify token in invoke payload and return AC response
+    * @param {authentication} authentication
+    */
+    async createAdaptiveCardInvokeResponseAsync(authentication) {
+        var auth_Obj = {
+            isTokenPresent: false,
+            isStatePresent: false,
+            isBasicRefresh: false
+        }
+
+        // verify token is present or not
+        if (authentication) {
+            auth_Obj.isTokenPresent = true;
+            auth_Obj.isStatePresent = true;
```

Comment:
either state will be there or token inside authentication

Reply:
Yes, Token is there inside Authentication but not state.

Reply:
why are you setting isStatePresent  = true. state is not part of authentication, it is present in parallel to authentication

Reply:
![image](https://user-images.githubusercontent.com/12997454/218965467-861013e9-a29d-4f82-b647-ca39383968e4.png)

Reply:
![image](https://user-images.githubusercontent.com/12997454/218965565-611efe22-4136-41d9-bb3e-ca6cf5a47b97.png)